cpp/src/community/k_truss_impl.cuh

```diff
@@ -144,247 +82,51 @@ edge_t remove_overcompensating_edges(raft::handle_t const& handle,
-      auto dst                 = invalid_dst_first[idx];
-      auto dst_array_end_valid = valid_dst_first + num_valid_edges;
-
-      auto itr_lower_valid =
-        thrust::lower_bound(thrust::seq, valid_dst_first, dst_array_end_valid, dst);
-      auto idx_lower_valid = thrust::distance(
-        valid_dst_first,
-        itr_lower_valid);  // Need a binary search to find the begining of the range
-
-      auto invalid_end_dst = invalid_dst_first + num_invalid_edges;
-
-      auto itr_lower_invalid =
-        thrust::lower_bound(thrust::seq, invalid_dst_first, invalid_end_dst, dst);
-      auto idx_lower_invalid = thrust::distance(
-        invalid_dst_first,
-        itr_lower_invalid);  // Need a binary search to find the begining of the range
-
-      auto incoming_edges_to_r_first_valid = thrust::make_zip_iterator(
-        valid_src_first + idx_lower_valid, thrust::make_constant_iterator(dst));
-      thrust::copy(
-        thrust::seq,
-        incoming_edges_to_r_first_valid,
-        incoming_edges_to_r_first_valid + (prefix_sum_valid[idx + 1] - prefix_sum_valid[idx]),
-        incoming_edges_to_r + prefix_sum_valid[idx] + prefix_sum_invalid[idx]);
-
-      auto incoming_edges_to_r_first_invalid = thrust::make_zip_iterator(
-        invalid_src_first + idx_lower_invalid, thrust::make_constant_iterator(dst));
-      thrust::copy(
-        thrust::seq,
-        incoming_edges_to_r_first_invalid,
-        incoming_edges_to_r_first_invalid + (prefix_sum_invalid[idx + 1] - prefix_sum_invalid[idx]),
-
-        incoming_edges_to_r + prefix_sum_invalid[idx] + prefix_sum_valid[idx + 1]);
+template <typename vertex_t, typename edge_t>
+struct extract_edges {
+  __device__ thrust::optional<thrust::tuple<vertex_t, vertex_t, edge_t>> operator()(
+    
+    auto src, auto dst, thrust::nullopt_t, thrust::nullopt_t, auto count) const
+  {
+    return thrust::make_tuple(src, dst, count);
+  }
+};
 
-      if constexpr (is_q_r_edge) {
-        auto potential_closing_edges_first_valid = thrust::make_zip_iterator(
-          valid_src_first + idx_lower_valid, thrust::make_constant_iterator(src));
-        thrust::copy(
-          thrust::seq,
-          potential_closing_edges_first_valid,
-          potential_closing_edges_first_valid + (prefix_sum_valid[idx + 1] - prefix_sum_valid[idx]),
-          potential_closing_edges + prefix_sum_valid[idx] + prefix_sum_invalid[idx]);
-
-        auto potential_closing_edges_first_invalid = thrust::make_zip_iterator(
-          invalid_src_first + idx_lower_invalid, thrust::make_constant_iterator(src));
-        thrust::copy(thrust::seq,
-                     potential_closing_edges_first_invalid,
-                     potential_closing_edges_first_invalid +
-                       (prefix_sum_invalid[idx + 1] - prefix_sum_invalid[idx]),
-                     potential_closing_edges + prefix_sum_invalid[idx] + prefix_sum_valid[idx + 1]);
+template <typename vertex_t>
+struct extract_edges_to_q_r {
```

Comment:
I feel like we have many functors and their names are a bit confusing.

In general, if we are using this functor in only one place of the code, using a lambda function might be more readable (unless we already know the functor such as thrust::plus or functor name really helps readability).

Historically, we defined functors using struct way more than we should in the past as nvcc had multiple issues in supporting lamda functors. It seems like most of those issues are resolved with cuda::proclaim_return_type. For simple functors (and if naming them doesn't help much in code readability), defining a functor closer to the caller site (i.e. using lambda) might be more readable. For functors used called only one place in the codebase, we may consider defining those in the caller site.

I will re-review all the non-lambda functors once we cleanup the code in the main-body of K-Truss.